File: chrome-extension/arhint-signer.js
const { spawnSync } = require('child_process');

function listCertificates() {
  const script = `Get-ChildItem Cert:\\CurrentUser\\My | Where-Object { $_.HasPrivateKey -and $_.NotAfter -gt (Get-Date) -and $_.NotBefore -lt (Get-Date) } | ForEach-Object { [PSCustomObject]@{ Thumbprint = $_.Thumbprint; Subject = $_.Subject; Issuer = $_.Issuer; NotBefore = $_.NotBefore.ToString('o'); NotAfter = $_.NotAfter.ToString('o'); HasPrivateKey = $_.HasPrivateKey; RawData = [Convert]::ToBase64String($_.RawData); FriendlyName = $_.FriendlyName } } | ConvertTo-Json`;
  
  const result = spawnSync('powershell.exe', [
    '-NoProfile',
    '-NonInteractive', 
    '-ExecutionPolicy', 'Bypass',
    '-Command', script
  ], { 
    encoding: 'utf8',
    maxBuffer: 10 * 1024 * 1024,
    windowsHide: true
  });
  
  if (result.error) {
    console.error('Failed to execute PowerShell:', result.error.message);
    return [];
  }
  
  if (result.stderr && result.stderr.trim()) {
    console.error('PowerShell stderr:', result.stderr);
  }
  
  if (!result.stdout || !result.stdout.trim()) {
    return [];
  }
  
  try {
    const parsed = JSON.parse(result.stdout);
    const certArray = Array.isArray(parsed) ? parsed : [parsed];
    
    return certArray.map(cert => {
      // Parse subject to extract friendly name
      const subject = cert.Subject;
      let displayName = '';
      
      // Try to extract Given Name and Surname first (for personal certs)
      const gnMatch = subject.match(/G=([^,+]+)/);
      const snMatch = subject.match(/SN=([^,+]+)/);
      
      if (gnMatch && snMatch) {
        displayName = `${gnMatch[1].trim()} ${snMatch[1].trim()}`;
      } else {
        // Otherwise use CN (Common Name)
        const cnMatch = subject.match(/CN=([^,]+)/);
        if (cnMatch) {
          displayName = cnMatch[1].trim();
        }
      }
      
      // If still no display name, fall back to first part of subject
      if (!displayName) {
        displayName = subject.split(',')[0].replace(/^CN=/, '').trim();
      }
      
      // Extract Organization (O)
      const orgMatch = subject.match(/\bO=([^,+]+)/);
      const organization = orgMatch ? orgMatch[1].trim() : '';
      
      // Add organization to display name if present
      if (organization) {
        displayName += ` (${organization})`;
      }
      
      // Extract issuer CN
      const issuer = cert.Issuer;
      const issuerCnMatch = issuer.match(/CN=([^,]+)/);
      const issuerName = issuerCnMatch ? issuerCnMatch[1].trim() : issuer.split(',')[0].trim();
      
      // Add expiry date to label for context
      const expiryDate = new Date(cert.NotAfter);
      const expiry = expiryDate.toLocaleDateString();
      
      return {
        label: `Issued for: ${displayName} | Issuer: ${issuerName} (expires ${expiry})`,
        thumbprint: cert.Thumbprint,
        subject: cert.Subject,
        issuer: cert.Issuer,
        notBefore: cert.NotBefore,
        notAfter: cert.NotAfter,
        hasPrivateKey: cert.HasPrivateKey,
        cert: cert.RawData
      };
    });
  } catch (e) {
    console.error('Failed to parse certificate data:', e.message);
    return [];
  }
}

function signHash(hashB64, thumbprint) {
  // Validate inputs
  if (!hashB64 || typeof hashB64 !== 'string') {
    throw new Error('Hash is required and must be a string');
  }
  
  if (!thumbprint || typeof thumbprint !== 'string') {
    throw new Error('Thumbprint is required and must be a string');
  }
  
  // Validate base64 format
  const base64Regex = /^[A-Za-z0-9+/]*={0,2}$/;
  if (!base64Regex.test(hashB64)) {
    throw new Error('Hash must be valid base64 encoded string');
  }
  
  const script = `$cert = Get-ChildItem Cert:\\CurrentUser\\My | Where-Object { $_.Thumbprint -eq '${thumbprint}' }; if (-not $cert) { throw 'Certificate not found' }; if (-not $cert.HasPrivateKey) { throw 'Certificate has no private key' }; $hashBytes = [Convert]::FromBase64String('${hashB64}'); $rsa = [System.Security.Cryptography.X509Certificates.RSACertificateExtensions]::GetRSAPrivateKey($cert); $signature = $rsa.SignHash($hashBytes, [System.Security.Cryptography.HashAlgorithmName]::SHA256, [System.Security.Cryptography.RSASignaturePadding]::Pkcs1); [Convert]::ToBase64String($signature)`;
  
  const result = spawnSync('powershell.exe', [
    '-NoProfile',
    '-NonInteractive',
    '-ExecutionPolicy', 'Bypass',
    '-Command', script
  ], { 
    encoding: 'utf8',
    windowsHide: true
  });
  
  if (result.error) {
    throw new Error(`Failed to execute PowerShell: ${result.error.message}`);
  }
  
  if (result.stderr && result.stderr.trim()) {
    throw new Error(`Signing failed: ${result.stderr.trim()}`);
  }
  
  return result.stdout.trim();
}

// Chrome Native Messaging protocol helper functions
function readMessage(callback) {
  let chunks = [];
  let totalLength = 0;
  let messageLength = null;

  process.stdin.on('readable', () => {
    let chunk;
    while ((chunk = process.stdin.read()) !== null) {
      chunks.push(chunk);
      totalLength += chunk.length;

      // Read message length (first 4 bytes)
      if (messageLength === null && totalLength >= 4) {
        const buffer = Buffer.concat(chunks);
        messageLength = buffer.readUInt32LE(0);
        chunks = [buffer.slice(4)];
        totalLength -= 4;
      }

      // Read complete message
      if (messageLength !== null && totalLength >= messageLength) {
        const buffer = Buffer.concat(chunks);
        const message = buffer.slice(0, messageLength).toString('utf8');
        chunks = [buffer.slice(messageLength)];
        totalLength -= messageLength;
        messageLength = null;

        callback(JSON.parse(message));
      }
    }
  });
}

function writeMessage(message) {
  const buffer = Buffer.from(JSON.stringify(message), 'utf8');
  const header = Buffer.alloc(4);
  header.writeUInt32LE(buffer.length, 0);
  process.stdout.write(header);
  process.stdout.write(buffer);
}

// Handle incoming messages
readMessage(msg => {
  console.error("Received message:", msg);

  try {
    let result;
    
    if (msg.action === "listCerts") {
      result = listCertificates();
    } else if (msg.action === "sign") {
      result = signHash(msg.hash, msg.thumbprint);
    }
    
    writeMessage({ result });
  } catch (e) {
    console.error("Error:", e.message);
    writeMessage({ error: e.message });
  }
});

File: chrome-extension/arhint-signer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Security.Cryptography;
using System.Security.Cryptography.X509Certificates;
using System.Text;
using System.Text.RegularExpressions;
using System.Web.Script.Serialization;

namespace ArhintSigner
{
    class Program
    {
        static void Main(string[] args)
        {
            try
            {
                ReadMessage(HandleMessage);
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine("Fatal error: " + ex.Message);
                Environment.Exit(1);
            }
        }

        static void HandleMessage(Dictionary<string, object> msg)
        {
            Console.Error.WriteLine("Received message: " + new JavaScriptSerializer().Serialize(msg));

            try
            {
                object result = null;

                if (msg.ContainsKey("action"))
                {
                    string action = msg["action"].ToString();

                    if (action == "listCerts")
                    {
                        result = ListCertificates();
                    }
                    else if (action == "sign")
                    {
                        if (!msg.ContainsKey("hash") || !msg.ContainsKey("thumbprint"))
                        {
                            throw new Exception("Missing required parameters for sign action");
                        }

                        result = SignHash(msg["hash"].ToString(), msg["thumbprint"].ToString());
                    }
                    else
                    {
                        throw new Exception("Unknown action: " + action);
                    }
                }

                WriteMessage(new { result });
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine("Error: " + ex.Message);
                WriteMessage(new { error = ex.Message });
            }
        }

        static List<object> ListCertificates()
        {
            var results = new List<object>();

            using (var store = new X509Store(StoreName.My, StoreLocation.CurrentUser))
            {
                store.Open(OpenFlags.ReadOnly);

                var now = DateTime.Now;

                foreach (var cert in store.Certificates)
                {
                    // Filter: must have private key and be currently valid
                    if (cert.HasPrivateKey && cert.NotAfter > now && cert.NotBefore < now)
                    {
                        try
                        {
                            // Parse subject to extract friendly name
                            string subject = cert.Subject;
                            string displayName = "";

                            // Try to extract Given Name and Surname first (for personal certs)
                            var gnMatch = Regex.Match(subject, @"G=([^,+]+)");
                            var snMatch = Regex.Match(subject, @"SN=([^,+]+)");

                            if (gnMatch.Success && snMatch.Success)
                            {
                                displayName = gnMatch.Groups[1].Value.Trim() + " " + snMatch.Groups[1].Value.Trim();
                            }
                            else
                            {
                                // Otherwise use CN (Common Name)
                                var cnMatch = Regex.Match(subject, @"CN=([^,]+)");
                                if (cnMatch.Success)
                                {
                                    displayName = cnMatch.Groups[1].Value.Trim();
                                }
                            }

                            // If still no display name, fall back to first part of subject
                            if (string.IsNullOrEmpty(displayName))
                            {
                                displayName = subject.Split(',')[0].Replace("CN=", "").Trim();
                            }

                            // Extract Organization (O)
                            var orgMatch = Regex.Match(subject, @"\bO=([^,+]+)");
                            string organization = orgMatch.Success ? orgMatch.Groups[1].Value.Trim() : "";

                            // Add organization to display name if present
                            if (!string.IsNullOrEmpty(organization))
                            {
                                displayName += " (" + organization + ")";
                            }

                            // Extract issuer CN
                            string issuer = cert.Issuer;
                            var issuerCnMatch = Regex.Match(issuer, @"CN=([^,]+)");
                            string issuerName = issuerCnMatch.Success 
                                ? issuerCnMatch.Groups[1].Value.Trim() 
                                : issuer.Split(',')[0].Trim();

                            // Add expiry date to label for context
                            string expiry = cert.NotAfter.ToShortDateString();

                            results.Add(new
                            {
                                label = "Issued for: " + displayName + " | Issuer: " + issuerName + " (expires " + expiry + ")",
                                thumbprint = cert.Thumbprint,
                                subject = cert.Subject,
                                issuer = cert.Issuer,
                                notBefore = cert.NotBefore.ToString("o"),
                                notAfter = cert.NotAfter.ToString("o"),
                                hasPrivateKey = cert.HasPrivateKey,
                                cert = Convert.ToBase64String(cert.RawData)
                            });
                        }
                        catch (Exception ex)
                        {
                            Console.Error.WriteLine("Error processing certificate " + cert.Thumbprint + ": " + ex.Message);
                        }
                    }
                }

                store.Close();
            }

            return results;
        }

        static string SignHash(string hashB64, string thumbprint)
        {
            // Validate inputs
            if (string.IsNullOrEmpty(hashB64))
            {
                throw new Exception("Hash is required and must be a string");
            }

            if (string.IsNullOrEmpty(thumbprint))
            {
                throw new Exception("Thumbprint is required and must be a string");
            }

            // Validate base64 format
            if (!Regex.IsMatch(hashB64, @"^[A-Za-z0-9+/]*={0,2}$"))
            {
                throw new Exception("Hash must be valid base64 encoded string");
            }

            byte[] hashBytes;
            try
            {
                hashBytes = Convert.FromBase64String(hashB64);
            }
            catch (Exception ex)
            {
                throw new Exception("Invalid base64 hash: " + ex.Message);
            }

            using (var store = new X509Store(StoreName.My, StoreLocation.CurrentUser))
            {
                store.Open(OpenFlags.ReadOnly);

                var cert = store.Certificates
                    .Find(X509FindType.FindByThumbprint, thumbprint, false)
                    .Cast<X509Certificate2>()
                    .FirstOrDefault();

                if (cert == null)
                {
                    throw new Exception("Certificate not found");
                }

                if (!cert.HasPrivateKey)
                {
                    throw new Exception("Certificate has no private key");
                }

                try
                {
                    using (RSA rsa = cert.GetRSAPrivateKey())
                    {
                        if (rsa == null)
                        {
                            throw new Exception("Unable to get RSA private key from certificate");
                        }

                        byte[] signature = rsa.SignHash(hashBytes, HashAlgorithmName.SHA256, RSASignaturePadding.Pkcs1);
                        return Convert.ToBase64String(signature);
                    }
                }
                finally
                {
                    store.Close();
                }
            }
        }

        // Chrome Native Messaging protocol implementation
        static void ReadMessage(Action<Dictionary<string, object>> callback)
        {
            using (var stdin = Console.OpenStandardInput())
            using (var reader = new BinaryReader(stdin))
            {
                while (true)
                {
                    try
                    {
                        // Read message length (4 bytes, little-endian)
                        byte[] lengthBytes = reader.ReadBytes(4);
                        if (lengthBytes.Length < 4)
                        {
                            // EOF reached
                            return;
                        }

                        int messageLength = BitConverter.ToInt32(lengthBytes, 0);

                        if (messageLength <= 0 || messageLength > 10 * 1024 * 1024)
                        {
                            throw new Exception("Invalid message length: " + messageLength);
                        }

                        // Read the message
                        byte[] messageBytes = reader.ReadBytes(messageLength);
                        if (messageBytes.Length < messageLength)
                        {
                            throw new Exception("Incomplete message received");
                        }

                        string messageJson = Encoding.UTF8.GetString(messageBytes);
                        JavaScriptSerializer serializer = new JavaScriptSerializer();
                        Dictionary<string, object> message = serializer.Deserialize<Dictionary<string, object>>(messageJson);

                        callback(message);
                    }
                    catch (EndOfStreamException)
                    {
                        // Normal exit when stdin is closed
                        return;
                    }
                }
            }
        }

        static void WriteMessage(object message)
        {
            JavaScriptSerializer serializer = new JavaScriptSerializer();
            string json = serializer.Serialize(message);
            byte[] messageBytes = Encoding.UTF8.GetBytes(json);
            byte[] lengthBytes = BitConverter.GetBytes(messageBytes.Length);

            using (var stdout = Console.OpenStandardOutput())
            using (var writer = new BinaryWriter(stdout))
            {
                writer.Write(lengthBytes);
                writer.Write(messageBytes);
                writer.Flush();
            }
        }
    }
}
